### PR TITLE
[IE MyriadX] - LSTMSequence - 3 output

### PR DESCRIPTION
**Pull request by DariaMityagina**
**-26693** - Adding the possibility of obtaining the 3rd output (LSTMSequence)
mdk PR: 14214

### CODE REVIEW

**Comment by DariaMityagina** [comment]
@andrejsokolov
could you please take a look?
work in progress.

File: inference-engine/src/legacy_api/src/net_pass.cpp [hunk 1]
```diff
@@ -304,6 +304,19 @@ void RemoveLayer(CNNLayerPtr& layer) {
 /****  Converter Passes  ************************************/
 /************************************************************/
 
+static RNNSequenceLayer::CellType cell_type_from_name(std::string& layer_type) {
```
Comment (andrejsokolov) [inline inference-engine/src/legacy_api/src/net_pass.cpp:307]: @DariaMityagina, this file does not belong to our plugin. Let us leave it without modification. It is unnecessary.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 2]
```diff
@@ -138,6 +143,18 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
 
     size_t stateSize = outputs[0]->desc().totalDimSize() / nCells / nBatches;
     size_t cellStateSize = inputs[2]->desc().totalDimSize() / nBatches;
+
+    if (outputs.size() == 3) {
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:147]: why are you trying to calculate this parameter by this formula?  stateSize should be the same.
It seems in DeepSpeech nCells and nBatches are swaped.
You can add:
`
    size_t nCells = inputs[0]->desc().dim(Dim::C);
    size_t nBatches = inputs[0]->desc().dim(Dim::H);
`
as a temporary solution when number of outputs == 3.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 3]
```diff
@@ -132,16 +132,27 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
 
     size_t nCells = inputs[0]->desc().dim(Dim::H);
     size_t nBatches = inputs[0]->desc().dim(Dim::C);
+    size_t inputSize = inputs[0]->desc().dim(Dim::W);
 
-    if (outputs.size() == 3) {
-        nCells = inputs[0]->desc().dim(Dim::C);
+    int axis = 1; // TODO : get from xml
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:137]: take a look at GPU plugin implementation:
inference-engine/src/cldnn_engine/cldnn_lstm.cpp

they process this parameter as:
`    bool permute_input = (1 != rnnLayer->axis);
`
...

        if (!permute_input) {
            lstm_batch_size = in_dims0.front();
            lstm_sequence_len = in_dims0[1];
        } else {
            lstm_batch_size = in_dims0[1];
            lstm_sequence_len = in_dims0.front();
        }
`
you can do similar thing.

File: inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.cpp [hunk 4]
```diff
@@ -165,13 +185,22 @@ TEST_F(myriadLayersTests_nightly, LSTMCellSequenceNet) {
...
+        InferenceEngine::Blob::Ptr output;
+        ASSERT_NO_THROW(st = inferRequest->GetBlob("RNNOutput", output, &_resp));
+        ASSERT_EQ(StatusCode::OK, st) << _resp.msg;
+        CompareCommonAbsolute(output, refOut0, ERROR_BOUND);
+    }
 }
 
-INSTANTIATE_TEST_CASE_P(accuracy, myriadLayersTestsLSTMCell_smoke,
-                        ::testing::Values<lstmcell_test_params>(MAKE_STRUCT(lstmcell_test_params, 512, 128)),
+// INSTANTIATE_TEST_CASE_P(accuracy, myriadLayersTestsLSTMCell_nightly,
```
Comment (andrejsokolov) [inline inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.cpp:196]: dead code?

File: inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.cpp [hunk 5]
```diff
@@ -3,12 +3,15 @@
 //
 
 #include "myriad_layers_lstm_cell.hpp"
+#include "blob_factory.hpp"
 
 TEST_F(myriadLayersTests_nightly, LSTMCellSequenceNet) {
```
Comment (andrejsokolov) [inline inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.cpp:8]: Please do not remove old test with previous IR. Just add one more with new IR.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 6]
```diff
@@ -130,14 +132,31 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
 
     size_t nCells = inputs[0]->desc().dim(Dim::H);
     size_t nBatches = inputs[0]->desc().dim(Dim::C);
+    size_t inputSize = inputs[0]->desc().dim(Dim::W);
+
+    bool permute_input = (1 != layer->axis);
+
+    if (!permute_input) {
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:139]: make this condition more simple: process each `axis` case or just permute_input/!permute_input (like in GPU plugin). Please do not mix them together.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 7]
```diff
@@ -158,35 +174,73 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
...
+        if (nCells > 1)
+            model->addTempBuffer(stage, DataDesc({stateSize}));
+
+        bool RNNForward = layer->direction == ie::RNNSequenceLayer::FWD;
+        stage->attrs().set<bool>("RNNForward", RNNForward);
+        stage->attrs().set<int>("nCells", nCells);
+        stage->attrs().set<int>("nBatches", nBatches);
+    }
+
+    if (outputs.size() == 3) {
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:223]: the difference between ouputs.size() == 2 and outputs.size()==3 is only outputs? So why do you duplicate the other code? you can create ouputs and fill it in accordance with outputs number.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 8]
```diff
@@ -61,14 +61,15 @@ class LSTMCellStage final : public StageNode {
...
+        IE_ASSERT(numOutputs() < 4);
         assertAllInputsOutputsTypes(this, DataType::FP16, DataType::FP16);
     }
 
     void serializeParamsImpl(BlobSerializer& serializer) const override {
         const auto RNNForward = attrs().get<bool>("RNNForward");
         const auto nCells     = attrs().get<int>("nCells");
         const auto nBatches   = attrs().get<int>("nBatches");
-        const bool useCellState = outputEdges().size() >= 2;
+        const bool useCellState = outputEdges().size() >= 1;
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:72]: why do you need to change this check?

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 9]
```diff
@@ -158,35 +174,73 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
...
+        stage->attrs().set<bool>("RNNForward", RNNForward);
+        stage->attrs().set<int>("nCells", nCells);
+        stage->attrs().set<int>("nBatches", nBatches);
+    }
 }
 
 void FrontEnd::parseLSTMCell(const Model& model, const ie::CNNLayerPtr& _layer, const DataVector &inputs, const DataVector &outputs) {
     IE_ASSERT(inputs.size() == 3);
-    IE_ASSERT(outputs.size() == 2);
+    IE_ASSERT(outputs.size() <= 3);
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:243]: it is necessary?

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp (hunk 8 above)
Comment (DariaMityagina) [reply inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:72]: well, i don't

File: inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.cpp [hunk 10]
```diff
@@ -30,15 +75,22 @@ TEST_F(myriadLayersTests_nightly, LSTMCellSequenceNet) {
...
     }
-    ie_fp16 *bias = weights0 + num_weights;
+
     for (int i = 0; i < num_bias; i++) {
         bias[i] = PrecisionUtils::f32tof16((float)((rand() % num_bias)) / num_bias);
         *(weights_for_net + num_weights + i) = bias[i];
     }
 
     InferenceEngine::Core ie;
     auto full_network = ie.ReadNetwork(tensorIteratorModel, weightsBlob_for_net);
```
Comment (andrejsokolov) [inline inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.cpp:85]: put this `ReadNetwork` under 2nd condition.

**Comment by andrejsokolov** [comment]
@DariaMityagina, do you serialize `outputsNumber` in parseRNN? I did not find it. On the device part, you are reading it from Blob.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp (hunk 8 above)
Comment (andrejsokolov) [reply inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp]: let us remove it

**Comment by DariaMityagina** [comment]
> @DariaMityagina, do you serialize `outputsNumber` in parseRNN? I did not find it. On the device part, you are reading it from Blob.

No, i use `outputs.size()` for getting outputs number.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 11]
```diff
@@ -98,6 +99,7 @@ class LSTMCellStage final : public StageNode {
 
 static void RNNRelayout(
```
Comment (avershov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:100]: Why it's static? We can use anonimous namespace for local entities

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 12]
```diff
@@ -102,18 +102,21 @@ static void RNNRelayout(
 
                  fp16_t* dst0,
                  fp16_t* dst1,
+                 fp16_t* dst2,
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:105]: why do you change this method? Passing number of outputs to Blob means that you should add one more parameter to the serializer.

Comment (andrejsokolov) [reply inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:105]: like `serializer.append(static_cast<int>(useCellState))`

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp (hunk 11 above)
Comment (andrejsokolov) [reply inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:100]: Actually, there should not be changes in this method at all. Let us leave it without changes for now.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 13]
```diff
@@ -198,19 +252,23 @@ void FrontEnd::parseLSTMCell(const Model& model, const ie::CNNLayerPtr& _layer,
 
     const auto& src = inputs[0]->desc();
 
-    const std::size_t nBatches = src.dim(Dim::N);
+    std::size_t nBatches = src.dim(Dim::N);
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:255]: Why not `const`. It is better to declare variable as `const` if they are not supposed to be modified.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 14]
```diff
@@ -130,14 +133,28 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
 
     size_t nCells = inputs[0]->desc().dim(Dim::H);
```
Comment (andrejsokolov) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:134]: size_t nCells = 0;
size_t nBatches = 0;

**Comment by andrejsokolov** [comment]
@avershov, do you have any comments?

**Comment by andrejsokolov** [comment]
@gladilov-gleb, please take a look. If there is no something serious, let us merge it.

File: inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp [hunk 15]
```diff
@@ -158,22 +175,37 @@ void FrontEnd::parseRNN(const Model& model, const ie::CNNLayerPtr& _layer, const
...
+        outputData.push_back(outputs[0]);
+        outputData.push_back(stateCellFinal);
+    } else if (outputs.size() == 2) {
+        outputData.push_back(outputs[0]);
+        outputData.push_back(outputs[1]);
+    } else if (outputs.size() == 3) {
+        outputData.push_back(outputs[0]);
+        outputData.push_back(outputs[1]);
+        outputData.push_back(outputs[2]);
+    }
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/graph_transformer/src/stages/rnn.cpp:201]: Nitpick: I'd suggest another form:

```cpp
auto outputData = outputs;
// should there be outputs.size() < 3 ??
if (outputs.size() == 1) {
    outputData.push_back(model->addFakeData());
}
```

File: inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.hpp [hunk 16]
```diff
@@ -153,9 +153,204 @@ std::string tensorIteratorModel = R"V0G0N(
 </net>
 )V0G0N";
 
+std::string tensorIteratorModel_3 = R"V0G0N(
```
Comment (gladilov-gleb) [inline inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.hpp:156]: I'm not sure it's allowed to add new tests with hardcoded model, even in deprecated infra. Are there any other similar tests. My vision is if test needs to have hardcoded model it should be introduced in closed tests repo.

Comment (andrejsokolov) [reply inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.hpp:156]: it is mostly artificial IR. And there is no similar other tests. If you have doubts, we can remove this test for now.

Comment (andrejsokolov) [reply inference-engine/tests_deprecated/functional/vpu/common/layers/myriad_layers_lstm_cell.hpp:156]: But I guess it is ok.

**Comment by avershov** [comment]
@DariaMityagina For the future: using of 'rebase' is much preferable than 'merge' to update your work with the latest target branch state. With merge, your branch will contain _your_ commits, and then a lot of other commits on top of yours, so comparison of HEAD vs HEAD~N will contain a myriads of extraneous files & changes. With rebase all your commits will be on top of target branch, so comparison of HEAD vs HEAD~N will show me only _your_ changes.

**Comment by DariaMityagina** [comment]
> @DariaMityagina For the future: using of 'rebase' is much preferable than 'merge' to update your work with the latest target branch state. With merge, your branch will contain _your_ commits, and then a lot of other commits on top of yours, so comparison of HEAD vs HEAD~N will contain a myriads of extraneous files & changes - which of them are just yours? With rebase all your commits will be on top of target branch, so comparison of HEAD vs HEAD-N will show me only _your_ changes.

@avershov , thank you! I got it.

**Comment by DariaMityagina** [comment]
Closing this PR because main part of changes is already in master branch, I will bring tests in another PR soon.

**Pull request closed without merging**